### ISSUE TITLE
Package couch2pg for gardener

### ISSUE THREAD

**Issue by btbonval** [medic/medic-webapp#1800]
I don't know what this entails, but it should happen.

Gardener will launch the importer and ensure it continues to run.

**Comment by btbonval** [comment]
Clearly this depends on #1799

**Comment by btbonval** [comment]
Once that is done, as a follow-up task, medic-os must be setup to configure/launch the couch2pg adapter gardener package.

**Comment by btbonval** [comment]
couch2pg relies on environment variables for configuration. Either Gardener or some other device will be necessary to ensure couch2pg is run with the correct environment settings.

**Comment by btbonval** [comment]
I will need the help of @mandric or someone else familiar with gardener.

**Comment by btbonval** [comment]
step 1 package as npm.

**Comment by btbonval** [comment]
This is a useful resource for making npm packages.
http://www.anupshinde.com/posts/how-to-create-nodejs-npm-package/

Also don't forget `npm link` for installing local before uploading to the main NPM repo.

Need to check on naming conventions for medic projects pushed to npm.

**Comment by btbonval** [comment]
There is some major work in progress under https://github.com/medic/medic-analytics/issues/11 in its own branch.

I need to move files around to follow npm conventions. Moving files around in master will create a merge disaster later.

Let's see if I can knock out the issue above before moving files around for this issue.

**Comment by btbonval** [comment]
#1799 is done as is issue 11 referenced above. This issue should be ready for me to work on now.

**Comment by btbonval** [comment]
http://stackoverflow.com/questions/5178334/folder-structure-for-a-node-js-project
more folder layout advice for node projects

I'm not sure about splitting my integrations into `/spec` and units into `/tests`, but maybe it'll work fine.

**Comment by btbonval** [comment]
Probably better to put this off until the analytics software is done.

Still need https://github.com/medic/medic-webapp/issues/1798 before the software is really ready for 0.4 and 0.6 production.

**Comment by btbonval** [comment]
It looks like this is going to be two mildly independent pieces of software under one hood. it will be useful to perform a directory restructure now which suits this packaging effort and also supports clear distinctions between import and form handling.

**Comment by btbonval** [comment]
Layouts updated in commit https://github.com/medic/medic-analytics/commit/d4f70bdeac1f7dad14056952ec60f165859fd5e3

**Comment by btbonval** [comment]
currently adding contacts work into the adapter. Assuming nothing else pops up, I can package the software.

**Comment by btbonval** [comment]
Contacts work was completed.

Added top-level index.js loop couch2pg import and xml form processing. Referenced in package.json as main.

Published to npm as medic-couch2pg version 0.0.2

**Comment by btbonval** [comment]
Have to run through a bunch of gardener stuff now:
https://github.com/garden20/gardener

**Comment by mandric** [comment]
Basically just create an npm package. Then we [add a configuration](https://github.com/medic/medic-webapp/blob/develop/kanso.json#L35) to `kanso.json` that tells [kanso-gardener](https://github.com/kanso/kanso-gardener) to package it up and attach to the ddoc.  kanso-gardener just runs `npm pack` on a directory and creates an attachment.

**Comment by mandric** [comment]
Think a nice-to-have is a published npm module but if the module is useless other than to our app you can just add a submodule to the medic-webapp repo, similar to api and sentinel.

**Comment by btbonval** [comment]
@mandric the couch2pg part might be useful to others. the xml forms parsing probably less so. The forms parsing part is probably only useful to us.

I was thinking I might split the two at some point. In preparation, I created two separate directories under libs, spec, and tests for the two separate pieces.

**Comment by btbonval** [comment]
Yeah I just looked at the link for gardener and all it says is "package it with npm" from what I can tell in the README.md. I should probably add the couchdb-worker tag.

**Comment by btbonval** [comment]
Env vars are a problem. The only env var exposed by the Kanso / Gardener / webapp process is COUCH_URL, and from that I'm meant to fetch all other configs.

I don't know anything about the process of putting configs into Couch directly. Considering these configs will be Henok, Milan, or tech lead work, I want to make sure I understand how those folks would normally manage these configs inside of Couch.

I looked into the API and Sentinel to see how those grabbed configs from Couch. Best I found was this, though I don't fully understand it.
https://github.com/medic/medic-sentinel/blob/master/db.js

**Comment by btbonval** [comment]
@browndav suggests moving most of the configs into postgresql itself, rather than couch. That is very easy for me.

The remaining pain point is the actual postgresql `postgres://user:password@site:port/dbname`, site:port are easy. dbname can be some default. user/pass might be read off a file on disk somewhere, still figuring that piece out.

For reference, the current configs:
https://github.com/medic/medic-analytics/#required-environment-variables

**Comment by btbonval** [comment]
nevermind the above, comment prior still stands unchanged.

**Comment by btbonval** [comment]
I need to write up the configs into the settings_schema of this file:
https://github.com/medic/medic-webapp/blob/develop/kanso.json#L83

As a note, we're not going to put PostgreSQL credentials into Couch. We're going to rely on a file on disk using unix security to protect it. Might as well link to the file location as a config in Couch because security through obscurity isn't reliable, we're relying on the unix perms to protect the file.

I don't know the contents of that file will look like. Ideally it'd be a `postgres://` URL with built-in user:pass creds.

Then the adapter itself can use app_settings garden tool to retrieve all these settings from Couch.

**Comment by SCdF** [comment]
Can I ask a dumb question: since this isn't a couchdb app, why would we use gardener? If we're just looking for a way of packaging and deploying something that is a little better than `git checkout <TAG>` would npm not be the right answer?

**Comment by mandric** [comment]
Gardener https://github.com/garden20/gardener allows us to update/deploy a new npm module along with the couchapp.  Currently we have no other way to deploy new npm packages (think services like sentinel or api) with a build of medic-webapp, other than gardener.  If medic-analytics doesn't require updates with new versions of medic-webapp then it can be managed independently of the webapp so doesn't need gardener.  

Adding medic-analytics to the build of webapp would be an easy tweak to kanso.json and adding a submodule to git, but I'm also pretty sure we don't want to continue bloating the ddoc this way.  Ideally whatever is attached to the ddoc would be very little, like a string that `npm install` supports would be ideal. Then gardener would just reference a link and call npm instead of using attached files to the ddoc. This wouldn't be hard, I think there's an issue open in the gardener repo.  I don't think gardener is the problem, I think it's a nice idea but needs minor refinement and can continue to work for us to deploy services the webapp depends on.

**Comment by SCdF** [comment]
> If medic-analytics doesn't require updates with new versions of medic-webapp then it can be managed independently of the webapp so doesn't need gardener.

We maybe should get on the Skypes or something and talk about deployments etc, cause maybe I'm missing a philosophy angle here, but no, medic-analytics doesn't necessarily need to be updated with new versions of medic-webapp: I'd argue for its complete independence honestly.

**Comment by garethbowen** [comment]
My preference is to use the kanso deployment method for now. It makes it easy to deploy and update as well as automatically keeping it running through gardener. It's also conceivable that it will be dependent on a specific version of the app if the schema changes dramatically.

I completely agree with @mandric that we can do better than bloating the ddoc, but I don't want to block this issue on that work.

#2206 will reduce the impact of bloating the ddoc.

**Comment by SCdF** [comment]
So right now the deployment approach is:
 - checking out the code to devsql
 - do a bunch of manual DB stuff
 - npm install
 - running the code in a tmux window 

My plan was to make the most minimal change possible to get the big code change (supporting edits and deletes, thus generally meaning it can sit there and run by itself without intervention) to be:
 - checking out the code to devsql
 - do much less manual db stuff
 - npm install
 - running a script that installs a init.d script that works as you'd expect

If we're hell-bent on using kanso now now now, someone who knows how this stuff works ( @mandric  / @garethbowen ) should have a chat to me and we can work out how to do this. Remember that this is being deployed (right now) to servers that don't actually have medic-webapp or similar deployed to them.

**Comment by garethbowen** [comment]
I forgot it's running on a separate server. @browndav Is the plan to run couch2pg on the medic-os instance or keep it separate?

We should probably leave the bundling decision until we've decided which host it should run on...

**Comment by browndav** [comment]
Short-term I think we'll need to maintain the current centralized-server "pull" setup, but I'd propose that we should have production 2.x servers running an automatically-maintained local replica by the end of next iteration (read: with Medic OS packages for Postgres), and eventually standardize on that across the board.

Maybe I'm missing something, but isn't @SCdF's `npm install` goal essentially a prerequisite for distributing the code via gardener anyway? I'd give a +1 to the `manual DB stuff + npm install` goal as a starting point, and we could hop on the Skypes tomorrow morning PDT to work out a way to automate the `manual db stuff` part (which I'd imagine might affect the OS-level packages one way or another).

**Comment by browndav** [comment]
Update: as a stepping stone, there's now a `medic-couch2pg` package in https://github.com/medic/medic-os. This was deadly easy, but there are concerns about multi-step upgrades, so I'm on the fence about whether or not we'd want to put it into production. Per a Slack conversation, @SCdF and I are going to try to reproduce the whole setup as an npm package for gardener, and either (a) succeed and use that, or (b) fail and file issues against gardener to try to succeed another time.

From Slack: "Maybe we modify the file permissions [on the PostgreSQL credentials file(s)] to admit gardener, switch over to the new `pouchdb-http` to keep npm from trying to build the native `sqlite3` module, add [a command-line] option to wait on (and then use) a password file for Postgres, and then package the whole thing up with the magic credentials file path specified. Do something sensible if the file never appears."

**Comment by browndav** [comment]
Decision from a Slack call this morning with @SCdF: create system accounts (either automatically, e.g. for Klipfolio, or from `medic-credentials` for actual people) and allow `trust` authentication for those specific system groups in `pg_hba.conf`. This seems like a reasonable alternative to credential-passing, and would all end up in the `medic-rdbms` (PostgreSQL) package. Planning to do this today.

To do: chat with @SCdF and @garethbowen about `gardener` limitations/instability and see if we can come up with a refactoring or replacement plan.

**Comment by garethbowen** [comment]
Decision from Hangouts with @browndav and @SCdF this morning.

1. Bundle medic-couch2pg as a gardener managed node server
2. Add the ability to disable medic-couch2pg through configuration which defaults to disabled so it doesn't waste cpu and disk space when not needed. The node server will stay running and monitoring the configuration even when disabled.
3. Investigate improvements to gardener or complete replacement as a separate issue.

**Comment by garethbowen** [comment]
Steps to packaging for gardener:

1. Use the method described by @browndav for getting DB credentials
2. Move all other environment variables and add an additional `enable` settings to the webapp app settings. Update couch2pg to check the webapp for settings and watch for settings. Example code for this is in sentinel and api in the `config.js` file.
3. Add couch2pg as a submodule of webapp.
4. Add couch2pg to the `node_module_folders` in the kanso.json of webapp
5. Add `cd couch2pg && npm install && cd ..` to webapp/scripts/ci/after_success.sh (or maybe improve this to use `npm --prefix` ?)

It should then just work.

**Comment by browndav** [comment]
FYI: I have a working system package for `medic-couch2pg` in `medic/medic-os` as of https://github.com/medic/medic-os/commit/36cd3d7003c638f60e79eddc11dc4e06f3f9cf58.

**Comment by garethbowen** [comment]
Deprioritised.

**Comment by nice-snek** [comment]
hi frand @sglangevin, U is coot

Please close or schedule before the end of this sprint. See [triaging old issues](https://github.com/medic/medic-docs/blob/master/md/dev/workflow.md#triaging-old-issues).

Nice Snek

**Comment by sglangevin** [comment]
@browndav is this still needed? If not, please close.

**Comment by browndav** [comment]
Thanks, closing.
